### PR TITLE
Adding NFS CLI arg options and NFS port

### PR DESCRIPTION
**Pull request by nsarras**

### CODE REVIEW

File: dss_datamover/nfs_cluster.py [hunk 1]
```diff
@@ -41,7 +41,19 @@ class NFSCluster:
     manager = Manager()
 
     def __init__(self, config={}, user_id="ansible", password="password", logger=None):
-        self.config = config.get("nfs", {})
+        self.config = config.get("fs_config", {}).get("nfs", {})
+        self.nfs_port = config.get("fs_config", {}).get("nfsport", {})
```
Comment (grandsuri) [inline dss_datamover/nfs_cluster.py:45]: All this code should be part of overriding fs_config in master_application.py. Then the code doesn't need to be changed anywhere. 
You can move all of this code to a separate function in master_application.py.
Also there is a function validate_s3_prefix in master_application.py which expects the fs_config fields to validate the prefix. That code is broken.

File: dss_datamover/nfs_cluster.py [hunk 2]
```diff
@@ -41,7 +41,19 @@ class NFSCluster:
...
+        # override with CLI args if applicable
+        nfs_cli_args = ['nfs_server', 'nfs_port', 'nfs_share']
+        if set(nfs_cli_args).issubset(config):
+            if config['nfs_server'] is not None and config['nfs_share'] is not None:
```
Comment (grandsuri) [inline dss_datamover/nfs_cluster.py:49]: `is not None` is not needed. Just `if config['nfs_server'] and config['nfs_share']` is good enough.

File: dss_datamover/nfs_cluster.py (hunk 1 above)
Comment (grandsuri) [inline dss_datamover/nfs_cluster.py:45]: Default value for nfsport can be set here if not set in config.json or in the cli args

Comment (grandsuri) [inline dss_datamover/nfs_cluster.py:45]: self.nfs_port should be an integer, not a dict

File: dss_datamover/utils/config.py [hunk 3]
```diff
@@ -206,6 +206,12 @@ def put(self, subparser):
                                help='Profiling of PUT operation (Not Implemented)')
         subparser.add_argument("--standalone", "-sa", required=False, action='store_true',
                                help='Run in standalone mode')
+        subparser.add_argument("--nfs-server", "-ns", type=str, required='--nfs-port' in sys.argv or '--nfs-share' in sys.argv,
```
Comment (grandsuri) [inline dss_datamover/utils/config.py:209]: required should be False. This is just an option to override the configuration parameters. Completely optional
Same for the other 2 parameters as well

File: dss_datamover/utils/config.py [hunk 4]
```diff
@@ -255,6 +261,12 @@ def delete(self, subparser):
                                help='Run DataMover in debug mode')
         subparser.add_argument("--profile", "-pro", required=False, action='store_true',
                                help='Profiling of DEL operation (Not Implemented)')
+        subparser.add_argument("--nfs-server", "-ns", type=str, required='--nfs-port' in sys.argv or '--nfs-share' in sys.argv,
```
Comment (grandsuri) [inline dss_datamover/utils/config.py:264]: same comments as above

File: dss_datamover/utils/config.py [hunk 5]
```diff
@@ -255,6 +261,12 @@ def delete(self, subparser):
...
+                               help='specify NFS server ip addr')
+        subparser.add_argument("--nfs-port", "-np", type=str, required='--nfs-server' in sys.argv or '--nfs-share' in sys.argv,
+                               help='specify NFS server port')
+        subparser.add_argument("--nfs-share", "-nsh", type=str, required='--nfs-server' in sys.argv or '--nfs-port' in sys.argv,
+                               help='specify NFS share/directory to use')
```
Comment (grandsuri) [inline dss_datamover/utils/config.py:269]: We can make nfs-server and nfs-share as mutually dependent. If one is given, then the other one also has to be given

**Comment by grandsuri** [comment]
You can add code something like this

`
def check_nfs_share_param(argParse.action):
    def __call__(self, parser, namespace, values, option_string=None):
          if not namespace['nfs_server']:
                parser.error('Missing NFS Server IP')
         else:
               config['nfs_server'] = NFS_SERVER_IP


       subparser.add_argument("--nfs-server", "-ns", type=str, required=False, action=check_nfs_share_param, dest="nfs_server',
                               help='specify NFS server port')
        subparser.add_argument("--nfs-share", "-nsh", type=str, required=False, action=check_nfs_server_param, dest='nfs_share',
                               help='specify NFS share/directory to use')

`
You can do something like this to make it mutual dependent.

**Comment by nsarras** [comment]
> You can add code something like this
> 
> ```
> def check_nfs_share_param(argParse.action):
>     def __call__(self, parser, namespace, values, option_string=None):
>           if not namespace['nfs_server']:
>                 parser.error('Missing NFS Server IP')
>          else:
>                config['nfs_server'] = NFS_SERVER_IP
> 
> 
>        subparser.add_argument("--nfs-server", "-ns", type=str, required=False, action=check_nfs_share_param, dest="nfs_server', default=None,
>                                help='specify NFS server port')
>         subparser.add_argument("--nfs-share", "-nsh", type=str, required=False, action=check_nfs_server_param, dest='nfs_share', default=None
>                                help='specify NFS share/directory to use')
> ```
> 
> You can do something like this to make it mutual dependent.

With regards to this, if you see the way I am setting the args to required if the other arg shows up, I am already making it mutually dependent in a much simpler manner.

File: dss_datamover/master_application.py [hunk 6]
```diff
@@ -84,7 +84,15 @@ def __init__(self, operation, config):
         self.client_password = config["client"]["password"]
 
         self.s3_config = config.get("s3_storage", {})
+
         self.fs_config = self.config.get("fs_config", {})
+        # override NFS configs with CLI args if applicable
+        nfs_cli_args = ['nfs_server', 'nfs_port', 'nfs_share']
+        if set(nfs_cli_args).issubset(self.config):
```
Comment (grandsuri) [inline dss_datamover/master_application.py:91]: This line and the above one is not needed. We are blinding overwriting if the fields are given or add the nfs shares fields if not present in config.

File: dss_datamover/utils/config.py [hunk 7]
```diff
@@ -255,6 +261,12 @@ def delete(self, subparser):
                                help='Run DataMover in debug mode')
         subparser.add_argument("--profile", "-pro", required=False, action='store_true',
                                help='Profiling of DEL operation (Not Implemented)')
+        subparser.add_argument("--nfs-server", "-ns", type=str, required='--nfs-share' in sys.argv,
```
Comment (grandsuri) [inline dss_datamover/utils/config.py:264]: This required field doesn't work for all modes except command line arguments on console. It is ok for now

File: dss_datamover/master_application.py [hunk 8]
```diff
@@ -84,7 +84,13 @@ def __init__(self, operation, config):
         self.client_password = config["client"]["password"]
 
         self.s3_config = config.get("s3_storage", {})
+
         self.fs_config = self.config.get("fs_config", {})
+        # override NFS configs with CLI args if applicable
+        if self.config['nfs_server'] and self.config['nfs_share']:
```
Comment (grandsuri) [inline dss_datamover/master_application.py:90]: this fails if the user starts app without --nfs-server/--nfs-share options

**Review by grandsuri — APPROVED** [review]
Looks good

File: dss_datamover/master_application.py (hunk 8 above)
Comment (nsarras) [reply dss_datamover/master_application.py:90]: as mentioned internally, it does not, as these options are supplied as keys in self.config with None as values if no CLI options are passed in

**Pull request merged**